부산대 Android_이창욱_2주차_과제

ichanguk opened the pull request:
## 코드 작성하면서 어려웠던 점

1. MVVM 패턴, data binding을 처음 사용해봐서 교수님, 구글링, gpt 등 많이 도움 받고 공부해서 작성했습니다.

2. 하나의 액티비티에 리사이클러 뷰도 2개 사용하고 view model도 2개 사용했는데 fragment를 사용하면 코드가 더 좋아질지 궁금합니다.(+ 리사이클러 뷰 어댑터 클래스를 따로 파일을 만드는 게 나은 방식인지도 궁금합니다.)

## 중점적으로 리뷰해주셨으면 하는 부분

1. data binding을 더 사용해서 main activity의 코드를 더 줄일 수 있는 부분이 있는 지 궁금합니다.

2. 리사이클러 뷰 구현이 전체적으로 어색하지 않은 지 궁금합니다.

3. 저장된 검색어 목록에 데이터를 추가할 때 name이 중복된 경우를 생각해서 place의 id값을 참조해서 사용했는데 올바르게 사용했는지 궁금합니다.

## 결과 화면

![week2_1](https://github.com/kakao-tech-campus-2nd-step2/android-map-keyword/assets/102301656/d3d9956a-bc12-4578-822b-838858361d7f) ![week2_2](https://github.com/kakao-tech-campus-2nd-step2/android-map-keyword/assets/102301656/1bb2f5e7-993d-4afa-a957-d1486b50e05b) ![week2_3](https://github.com/kakao-tech-campus-2nd-step2/android-map-keyword/assets/102301656/a8795a71-3e96-43f4-a1fa-37dbf8078323)

ichanguk committed:
docs: update README.md

- repository 설명 추가
- 1단계 기능 리스트 추가

ichanguk committed:
design: add Base Search Layout

검색어 입력 및 검색 결과를 표시할 기본 레이아웃 구현

ichanguk committed:
feat: update Insert Data Feature

SQLite를 이용해 로컬 DB에 데이터를 저장하는 기능 구현

ichanguk committed:
docs: update README.md

2단계 기능 리스트 추가

ichanguk committed:
feat: merge PlaceDatabaseHelper into PlaceRepository

ichanguk committed:
feat: update Search Feature

- 검색어를 입력하면 검색 결과 목록이 표시되도록 함
- MVVM 패턴과 data binding을 사용함

ichanguk committed:
feat: update Save Search Word Feature

- 검색 결과 목록에서 항목을 누르면 검색어 저장 목록에 추가되도록 함.

ichanguk committed:
feat: update Delete Saved Search Word Feature

clear 버튼을 누르면 해당 저장된 검색어가 목록에서 삭제되도록 구현

ichanguk committed:
docs: update README.md

기능 리스트에 기능 추가

ichanguk committed:
feat: update Delete And Insert Existing Search Word Feature

검색 결과에서 선택한 검색어가 이미 저장된 목록에 있을 경우, 기존 검색어를 삭제하고 다시 추가하도록 구현

ichanguk committed:
feat: update Store And Delete Search Word Using PlaceId

이름이 같은 장소가 있을 수 있기 때문에 placeId를 참조하여 저장 및 삭제하도록 구현

ichanguk committed:
refactor: improve Code Readability And Remove Duplicate Code

bigstark commented on app/src/main/java/campus/tech/kakao/map/MainActivity.kt:
```diff
@@ -1,11 +1,309 @@
 class MainActivity : AppCompatActivity() {
+    private lateinit var binding: ActivityMainBinding
+    private val placeViewModel by lazy { ViewModelProvider(this)[PlaceViewModel::class.java] }
+    private val savedSearchWordViewModel by lazy { ViewModelProvider(this)[SavedSearchWordViewModel::class.java] }
```
by viewModels()

bigstark replied:
[activity-ktx](https://developer.android.com/jetpack/androidx/releases/activity)

bigstark replied:
멘토링 때에도 말씀드렸다시피 전략적인 선택을 하면 됩니다. 다만 지금 현재 화면의 요구사항이 굉장히 단순하기 때문에 저는 fragment 로 따로 분리하지는 않을 것 같아요. 확장요구에 대한 낌새가 보일 때 fragment 로 분리하는 작업을 진행할 것 같습니다.

또한 같은 Fragment 혹은 Activity 내에서 ViewModel 을 여러개 사용하는 것은 관점에 따라 좋을 수도 나쁠 수도 있습니다. 위에도 언급했다시피 이 화면 자체의 요구사항이 매우 간단하기 떄문에 ViewModel 을 나누기 보다 저는 하나의 ViewModel 로 진행할 것 같습니다 :)

bigstark commented on app/src/main/java/campus/tech/kakao/map/MainActivity.kt:
```diff
@@ -1,11 +1,309 @@
+                    )
+                }
+            }
+        binding.searchResultRecyclerView.adapter = ResultRecyclerViewAdapter(placeItemClickListener)
```
Memory Leak, Garbage Collection

bigstark commented on app/src/main/java/campus/tech/kakao/map/MainActivity.kt:
```diff
@@ -1,11 +1,309 @@
+         */
+        @SuppressLint("NotifyDataSetChanged")
+        fun setPlaces(places: List<Place>) {
+            searchResultList = places
```
[ListAdapter](https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter)

bigstark replied:
DiffCallabck

bigstark commented on app/src/main/java/campus/tech/kakao/map/data/PlaceRepository.kt:
```diff
@@ -0,0 +1,64 @@
+import android.database.sqlite.SQLiteOpenHelper
+import campus.tech.kakao.map.model.Place
+
+class PlaceRepository(context: Context) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {
```
[Repository pattern](https://developer.android.com/codelabs/basic-android-kotlin-training-repository-pattern?hl=ko#0)

bigstark commented on app/src/main/java/campus/tech/kakao/map/viewmodel/SavedSearchWordViewModel.kt:
```diff
@@ -0,0 +1,51 @@
+
+    fun deleteSearchWordById(searchWord: SavedSearchWord) {
+        viewModelScope.launch {
+            withContext(Dispatchers.IO) {
+                repository.deleteSearchWordById(searchWord.id)
+                val currentList = _savedSearchWords.value ?: emptyList()
+                _savedSearchWords.postValue(currentList - searchWord)
+            }
```
[코틀린 코루틴의 정석](https://product.kyobobook.co.kr/detail/S000212376884)

bigstark commented on app/src/main/java/campus/tech/kakao/map/MainActivity.kt:
```diff
@@ -1,11 +1,309 @@
+                placeViewModel.searchPlacesByCategory(searchText)
+            } else {
+                placeViewModel.searchPlacesByCategory("")
+            }
```
요거 굳이 if 문으로 나눌 필요가 없어 보이는 로직인 것 같아요. searchText 가 empty 라면 "" 일 테니까요!

bigstark commented on app/src/main/java/campus/tech/kakao/map/viewmodel/PlaceViewModel.kt:
```diff
@@ -0,0 +1,44 @@
+        _searchResults.value = emptyList()
+    }
+
+    fun searchPlacesByCategory(category: String) {
+        viewModelScope.launch {
+            withContext(Dispatchers.IO) {
+                val results = placeRepository.getPlacesByCategory(category)
+                _searchResults.postValue(results)
+            }
+        }
+    }
```
Dispatchers.IO 처리는 잘해주셨어요. 다만 withContext 로 처리하기 보다는 launch 시점에 context 를 설정하는 것이 어떨까요?

```suggestion
    fun searchPlacesByCategory(category: String) {
        viewModelScope.launch(Dispatchers.IO) {
            val results = placeRepository.getPlacesByCategory(category)
            _searchResults.postValue(results)
        }
    }
```

bigstark replied:
더불어 viewModelScope 은 launch 시에 아무것도 설정하지 않으면 어떤 쓰레드에서 동작되는지 확인해보시면 좋을 것 같아요!

bigstark commented on app/src/main/java/campus/tech/kakao/map/viewmodel/PlaceViewModel.kt:
```diff
@@ -0,0 +1,44 @@
+
+class PlaceViewModel(application: Application) : AndroidViewModel(application) {
+    private val placeRepository: PlaceRepository = PlaceRepository(application)
+    private val _searchResults = MutableLiveData<List<Place>>()
+
+    val searchResults: LiveData<List<Place>> get() = _searchResults
```
[Backing field](https://kotlinlang.org/docs/properties.html#backing-fields) 에 대해서 잘 이해하고 계시군요! Coroutine 을 사용중이면, Flow 를 사용하여 처리해보는건 어떨까요?

https://readystory.tistory.com/207

bigstark reviewed (changes requested):
> 저장된 검색어 목록에 데이터를 추가할 때 name이 중복된 경우를 생각해서 place의 id값을 참조해서 사용했는데 올바르게 사용했는지 궁금합니다.

name 이 primary 가 될 수 없다는 말씀이시죠? 잘하셨습니다 :)

ichanguk committed:
refactor: refactor Suggested Improvements from code review

- repository에 context를 직접적으로 넘겨주지 않게 변경
- 리사이클러 뷰에서 notifydatasetchanged() 사용하지 않도록 변경
